src/osd/PrimaryLogPG.cc

```diff
@@ -9976,6 +10022,278 @@ struct C_Flush : public Context {
+      if (refs.find(target) == refs.end()) {
+	continue;
+      }
+      C_SetDedupChunks *fin = new C_SetDedupChunks(this, soid, get_last_peering_reset(), p.first);
+      ceph_tid_t tid = refcount_manifest(soid, target, refcount_t::CREATE_OR_GET_REF, 
+			      fin, move(chunks[p.first]));
+      mop->chunks[target] = make_pair(p.first, p.second.length());
+      mop->num_chunks++;
+      mop->tids[p.first] = tid;
+      fin->tid = tid;
+      dout(10) << __func__ << " oid: " << soid << " tid: " << tid
+	      << " target: " << target << " offset: " << p.first
+	      << " length: " << p.second.length() << dendl;
+    }
+    cur_off += r;
+  }
+
+  if (mop->tids.size()) {
+    manifest_ops[soid] = mop;
+    manifest_ops[soid]->op = op;
+  } else {
+    // size == 0
+    return 0;
+  }
+
+  return -EINPROGRESS;
+}
+
+int PrimaryLogPG::do_cdc(const object_info_t& oi, uint64_t off, 
+			 std::map<uint64_t, chunk_info_t>& chunk_map,
+			 std::map<uint64_t, bufferlist>& chunks)
+{
+  uint64_t cur_off = off;
+  string chunk_algo = pool.info.get_dedup_chunk_algorithm_name();
+  int64_t chunk_size = pool.info.get_dedup_cdc_chunk_size();
+  uint64_t max_window_size = static_cast<uint64_t>(pool.info.get_dedup_cdc_window_size());
+  uint64_t total_length = 0;
+
+  std::unique_ptr<CDC> cdc = CDC::create(chunk_algo, cbits(chunk_size)-1);
+  if (!cdc) {
+    dout(0) << __func__ << " unrecognized chunk-algorithm " << dendl;
+    return -EINVAL;
+  }
+
+  if (cur_off >= oi.size) {
+    return -ERANGE;
+  }
+  bufferlist bl;
+  /**
+   * We disable EC pool as a base tier of distributed dedup.
+   * The reason why we disallow erasure code pool here is that the EC pool does not support objects_read_sync(). 
+   * Therefore, we should change the current implementation totally to make EC pool compatible. 
+   * As s result, we leave this as a future work.
+   */
+  int r = pgbackend->objects_read_sync(
+      oi.soid, cur_off, max_window_size, 0, &bl);
+  if (r < 0) {
+    dout(0) << __func__ << " read fail " << " offset: " << cur_off
+      << " len: " << max_window_size << " r: " << r << dendl;
+    return r;
+  }
+  if (bl.length() == 0) {
+    dout(0) << __func__ << " got 0 byte during chunking " << dendl;
+    return r;
+  }
+
+  dout(10) << __func__ << " oid: " << oi.soid << " len: " << bl.length() 
+	  << " oi.size: " << oi.size << " window_size: " << max_window_size 
+	  << " chunk_size: " << chunk_size << dendl;
+
+  vector<pair<uint64_t, uint64_t>> cdc_chunks;
+  cdc->calc_chunks(bl, &cdc_chunks);
+
+  // get fingerprint 
+  for (auto p : cdc_chunks) {
+    bufferlist chunk;
+    uint64_t object_offset = off + p.first;
+    chunk.substr_of(bl, p.first, p.second);
+    hobject_t target = get_fpoid_from_chunk(oi.soid, chunk);
+    chunks[object_offset] = move(chunk);
+    chunk_map[object_offset] = chunk_info_t(0, p.second, target);
+    total_length += p.second;
+  }
+  return total_length;
```

Comment:
Let's say the window size is 1MB and do_cdc is configured to emit chunks of size 16k - 32k.  If the second to last chunk is 1MB - 20k, the last chunk might not be on an actual chunk boundary.  calc_chunks is going to return it as a chunk anyway because it is the end of the buffer, but on a 4MB object, it would tend to prevent dup detection with other examples of the same string.  You could resolve this by rewinding total_length to the offset of the end of the last whole chunk, but for now I'd rather you read in the whole object here as you did before (but with a single call to object_read_sync),  call do_cdc on the whole thing, eliminate the loop in start_dedup, and eliminate the pool.info.get_dedup_cdc_window_size() configuration parameter entirely.